packages/lit-ssr/README.md

```diff
@@ -0,0 +1,158 @@
+# lit-ssr
+
+lit-html & LitElement SSR server
+
+## Setup
+
+This repo requires Node 14+ because of it's use of experimental VM modules for
+creating VM sandbox for executing the client-side code in. It also requires
+templates/elements be authored using pre-release versions of lit-html@2 and
+lit-element@3.
+
+## Usage
+
+### High-level API: render HTML pages
+
+The easiest way to get started using lit-ssr is to add the `renderHTMLFile`
+middleware from the `render-html-file` module to your server. This middleware
+will server-side render any `Lit` custom elements included in html files served
+in your application. It handles URLs to files with `.html` or bare paths with
+`accept: text/html` or `accept: *`. If a static file does not exist at the
+specified URL path, the `fallback` file will be rendered/served.
+
+Below is an example usage of the lit-ssr `renderHTMLFile` middleware:
+
+```js
+// Example server.js
+
+import * as path from 'path';
+import Koa from 'koa';
+import staticFiles from 'koa-static';
+import koaNodeResolve from 'koa-node-resolve';
+import {renderHTMLFile} from 'lit-ssr/lib/render-html-file.js';
+
+const {nodeResolve} = koaNodeResolve;
+
+const root = /* Root of your webapp */;
+
+(async() => {
+
+  const port = 8080;
+  const app = new Koa();
+  app.use(nodeResolve({root}));
+  app.use(renderHTMLFile({root, fallback: 'index.html'})); // Use `renderHTMLFile` plugin
+  app.use(staticFiles(root));
+  app.listen(port, () => {
+    console.log(`Server listening on port ${port}`);
+  });
+
+})();
+```
+
+When using this middleware, there are a few steps you should do to prepare you
+HTML files for proper server rendering:
+
+- Add the `template-shadowroot` polyfill to hydrate server-rendered `<template shadowroot>` elements. For example:
+  ```html
+  <script type="module">
+    import {hydrateShadowRoots} from './node_modules/template-shadowroot/template-shadowroot.js';
+    hydrateShadowRoots(document.body);
+  </script>
+  ```
+- Add `ssr` attribute to scripts that _should_ be executed on the server, for
+  example, scripts that include custom element definitions. For example:
+  ```html
+  <script type="module" src="my-app-components.js" ssr></script>
```

Comment:
We need to have a whole design session around the ways to package up the low-level API's into things that users actually consume.

AFAIK it's fairly novel to package up framework SSR as (almost) zero-config middleware that consumes otherwise static HTML pages, which I thought was sorta neat. But it leaves open the question for how you initialize data on the page on the server in a declarative way (since the middleware is consuming .html files), so I went with this, but definitely arguable. There are a ton of options in this space, so we just need to explore them.